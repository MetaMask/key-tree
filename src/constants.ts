export const BYTES_KEY_LENGTH = 32;

export const MIN_BIP_44_DEPTH = 0;
export const MAX_BIP_44_DEPTH = 5;

export type MinBIP44Depth = typeof MIN_BIP_44_DEPTH;
export type MaxBIP44Depth = typeof MAX_BIP_44_DEPTH;
export type BIP44Depth = MinBIP44Depth | 1 | 2 | 3 | 4 | MaxBIP44Depth;

// BIP-44 derivation path:
// m / purpose' / coin_type' / account' / change / address_index
//
// Per BIP-43 / BIP-44, "purpose" should always be "44":
// m / 44' / coin_type' / account' / change / address_index
//
// The Ethereum "coin_type" is "60". Its "account" and "change" indices are
// always "0". Here's an example Ethereum HD path for account "0":
// m  / 44' / 60' / 0' / 0 / 0

export type AnonymizedBIP39Node = 'm';
export type BIP39Node = `bip39:${string}`;
export type HardenedBIP32Node = `bip32:${number}'`;
export type UnhardenedBIP32Node = `bip32:${number}`;
export type BIP32Node = HardenedBIP32Node | UnhardenedBIP32Node;

export const BIP44PurposeNodeToken = `bip32:44'`;

/**
 * e.g.
 * -  bip32:0
 * -  bip32:0'
 */
export const BIP_32_PATH_REGEX = /^bip32:\d+'?$/u;

/**
 * bip39:<SPACE_DELMITED_SEED_PHRASE>
 *
 * The seed phrase must consist of 12 <= 24 words.
 */
export const BIP_39_PATH_REGEX = /^bip39:([a-z]+){1}( [a-z]+){11,23}$/u;

export const BIP_32_HARDENED_OFFSET = 0x80000000;

export type HDPathString0 = AnonymizedBIP39Node;
export type HDPathString1 = `${HDPathString0} / ${HardenedBIP32Node}`;
export type HDPathString2 = `${HDPathString1} / ${HardenedBIP32Node}`;
export type HDPathString3 = `${HDPathString2} / ${HardenedBIP32Node}`;
export type HDPathString4 = `${HDPathString3} / ${BIP32Node}`;
export type HDPathString5 = `${HDPathString4} / ${BIP32Node}`;

export type CoinTypeHDPathString = HDPathString2;
export type ChangeHDPathString = HDPathString4;
export type AddressHDPathString = HDPathString5;

export type HDPathString =
  | HDPathString0
  | HDPathString1
  | HDPathString2
  | HDPathString3
  | HDPathString4
  | HDPathString5;

export type RootedHDPathTuple0 = readonly [BIP39Node];
export type RootedHDPathTuple1 = readonly [
  ...RootedHDPathTuple0,
  HardenedBIP32Node,
];
export type RootedHDPathTuple2 = readonly [
  ...RootedHDPathTuple1,
  HardenedBIP32Node,
];
export type RootedHDPathTuple3 = readonly [
  ...RootedHDPathTuple2,
  HardenedBIP32Node,
];
export type RootedHDPathTuple4 = readonly [...RootedHDPathTuple3, BIP32Node];
export type RootedHDPathTuple5 = readonly [...RootedHDPathTuple4, BIP32Node];

export type RootedHDPathTuple =
  | RootedHDPathTuple0
  | RootedHDPathTuple1
  | RootedHDPathTuple2
  | RootedHDPathTuple3
  | RootedHDPathTuple4
  | RootedHDPathTuple5;

export type PartialHDPathTuple1 = readonly [HardenedBIP32Node];
export type PartialHDPathTuple2 = readonly [
  ...PartialHDPathTuple1,
  HardenedBIP32Node,
];
export type PartialHDPathTuple3 = readonly [
  ...PartialHDPathTuple2,
  HardenedBIP32Node,
];
export type PartialHDPathTuple4 = readonly [...PartialHDPathTuple3, BIP32Node];
export type PartialHDPathTuple5 = readonly [...PartialHDPathTuple4, BIP32Node];
export type PartialHDPathTuple6 = readonly [BIP32Node];
export type PartialHDPathTuple7 = readonly [BIP32Node, BIP32Node];
export type PartialHDPathTuple8 = readonly [
  HardenedBIP32Node,
  BIP32Node,
  BIP32Node,
];
export type PartialHDPathTuple9 = readonly [HardenedBIP32Node, BIP32Node];
export type PartialHDPathTuple10 = readonly [
  HardenedBIP32Node,
  HardenedBIP32Node,
  BIP32Node,
];
export type PartialHDPathTuple11 = readonly [
  HardenedBIP32Node,
  HardenedBIP32Node,
  BIP32Node,
  BIP32Node,
];

export type CoinTypeToAddressTuple = PartialHDPathTuple8;

export type PartialHDPathTuple =
  | PartialHDPathTuple1
  | PartialHDPathTuple2
  | PartialHDPathTuple3
  | PartialHDPathTuple4
  | PartialHDPathTuple5
  | PartialHDPathTuple6
  | PartialHDPathTuple7
  | PartialHDPathTuple8
  | PartialHDPathTuple9
  | PartialHDPathTuple10
  | PartialHDPathTuple11;

/**
 * Every ordered subset of a full HD path tuple.
 */
export type HDPathTuple = RootedHDPathTuple | PartialHDPathTuple;

export type RootedSLIP10PathTuple = readonly [BIP39Node, ...BIP32Node[]];
export type SLIP10PathTuple = readonly BIP32Node[];
export type SLIP10Path = RootedSLIP10PathTuple | SLIP10PathTuple;

export type FullHDPathTuple = RootedHDPathTuple5;
