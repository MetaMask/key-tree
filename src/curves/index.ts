export * from './curve';
export * as secp256k1 from './secp256k1';
export * as ed25519 from './ed25519';
