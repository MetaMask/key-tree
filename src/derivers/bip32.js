// node
const crypto = require('crypto');
const assert = require('assert');
// npm
const secp256k1 = require('secp256k1');
const createKeccakHash = require('keccak');

const HARDENED_OFFSET = 0x80000000;

module.exports = {
  // standard
  deriveChildKey,
  // utility
  bip32PathToMultipath,
  privateKeyToEthAddress,
};

function bip32PathToMultipath(bip32Path) {
  let pathParts = bip32Path.trim().split('/');
  // strip "m" noop
  if (pathParts[0].toLowerCase() === 'm') {
    pathParts = pathParts.slice(1);
  }
  const multipath = pathParts.map((part) => `bip32:${part}`).join('/');
  return multipath;
}

function deriveChildKey(parentKey, pathPart) {
  const isHardened = pathPart.includes(`'`);
  const indexPart = pathPart.split(`'`)[0];
  const childIndex = parseInt(indexPart, 10);
  assert(childIndex < HARDENED_OFFSET, 'Invalid index');
  assert(parentKey.length === 64, 'Parent key invalid length');

  const parentPrivateKey = parentKey.slice(0, 32);
  const parentExtraEntropy = parentKey.slice(32);
  const secretExtension = deriveSecretExtension({ parentPrivateKey, childIndex, isHardened });

  const { privateKey, extraEntropy } = generateKey({
    parentPrivateKey,
    parentExtraEntropy,
    secretExtension,
  });

  return Buffer.concat([privateKey, extraEntropy]);
}

// the bip32 secret extension is created from the parent private or public key and the child index
function deriveSecretExtension({ parentPrivateKey, childIndex, isHardened }) {
  if (isHardened) {
    // Hardened child
    const indexBuffer = Buffer.allocUnsafe(4);
    indexBuffer.writeUInt32BE(childIndex + HARDENED_OFFSET, 0);
    const pk = parentPrivateKey;
    const zb = Buffer.alloc(1, 0);
    return Buffer.concat([zb, pk, indexBuffer]);
  }
  // Normal child
  const indexBuffer = Buffer.allocUnsafe(4);
  indexBuffer.writeUInt32BE(childIndex, 0);
  const parentPublicKey = secp256k1.publicKeyCreate(parentPrivateKey, true);
  return Buffer.concat([parentPublicKey, indexBuffer]);

}

function generateKey({ parentPrivateKey, parentExtraEntropy, secretExtension }) {
  const entropy = crypto.createHmac('sha512', parentExtraEntropy).update(secretExtension).digest();
  const keyMaterial = entropy.slice(0, 32);
  // extraEntropy is also called "chaincode"
  const extraEntropy = entropy.slice(32);
  const privateKey = secp256k1.privateKeyTweakAdd(parentPrivateKey, keyMaterial);
  return { privateKey, extraEntropy };
}

function keccak(a, bits = 256) {
  return createKeccakHash(`keccak${bits}`).update(a).digest();
}

function privateKeyToEthAddress(keyBuffer) {
  const privateKey = keyBuffer.slice(0, 32);
  const publicKey = secp256k1.publicKeyCreate(privateKey, false).slice(1);
  return keccak(publicKey).slice(-20);
}
