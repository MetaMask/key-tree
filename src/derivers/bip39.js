// node
const crypto = require('crypto');
// npm
const bip39 = require('bip39');

const ROOT_BASE_SECRET = Buffer.from('Bitcoin seed', 'utf8');

module.exports = {
  // standard
  deriveChildKey,
  // utility
  bip39MnemonicToMultipath,
};

function bip39MnemonicToMultipath(mnemonic) {
  return `bip39:${mnemonic.trim()}`;
}

// this creates a child key using bip39, ignoring the parent key
function deriveChildKey(_parentKey, pathPart) {
  const mnemonic = pathPart;
  const seedBuffer = bip39.mnemonicToSeed(mnemonic);
  const entropy = crypto.createHmac('sha512', ROOT_BASE_SECRET).update(seedBuffer).digest();

  return entropy;
}
