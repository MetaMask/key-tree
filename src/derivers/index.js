const bip32 = require('./bip32');
const bip39 = require('./bip39');

module.exports = {
  bip32,
  bip39,
};
