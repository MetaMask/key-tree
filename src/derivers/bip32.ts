import { assert } from '@metamask/utils';
import { keccak_256 as keccak256 } from '@noble/hashes/sha3';

import type { DeriveChildKeyArgs } from '.';
import { BYTES_KEY_LENGTH } from '../constants';
import { secp256k1 } from '../curves';
import type { SLIP10Node } from '../SLIP10Node';
import { isValidBytesKey, validateBIP32Index } from '../utils';
import type { DeriveNodeArgs } from './shared';
import {
  deriveChildKey as sharedDeriveChildKey,
  deriveSecretExtension,
  generateEntropy,
  derivePublicExtension,
} from './shared';

/**
 * Converts a BIP-32 private key to an Ethereum address.
 *
 * **WARNING:** Only validates that the key is non-zero and of the correct
 * length. It is the consumer's responsibility to ensure that the specified
 * key is a valid BIP-44 Ethereum `address_index` key.
 *
 * @param key - The `address_index` private key bytes to convert to an Ethereum
 * address.
 * @returns The Ethereum address corresponding to the given key.
 */
export function privateKeyToEthAddress(key: Uint8Array) {
  assert(
    key instanceof Uint8Array && isValidBytesKey(key, BYTES_KEY_LENGTH),
    'Invalid key: The key must be a 32-byte, non-zero Uint8Array.',
  );

  const publicKey = secp256k1.getPublicKey(key, false);
  return publicKeyToEthAddress(publicKey);
}

/**
 * Converts a BIP-32 public key to an Ethereum address.
 *
 * **WARNING:** Only validates that the key is non-zero and of the correct
 * length. It is the consumer's responsibility to ensure that the specified
 * key is a valid BIP-44 Ethereum `address_index` key.
 *
 * @param key - The `address_index` public key bytes to convert to an Ethereum
 * address.
 * @returns The Ethereum address corresponding to the given key.
 */
export function publicKeyToEthAddress(key: Uint8Array) {
  assert(
    key instanceof Uint8Array &&
      isValidBytesKey(key, secp256k1.publicKeyLength),
    'Invalid key: The key must be a 65-byte, non-zero Uint8Array.',
  );

  return keccak256(key.slice(1)).slice(-20);
}

/**
 * Derive a BIP-32 child key with a given path from a parent key.
 *
 * @param options - The options for deriving a child key.
 * @param options.path - The derivation path part to derive.
 * @param options.node - The node to derive from.
 * @param options.curve - The curve to use for derivation.
 * @returns The derived child key as a {@link SLIP10Node}.
 */
export async function deriveChildKey(
  options: DeriveChildKeyArgs,
): Promise<SLIP10Node> {
  assert(
    options.curve.name === 'secp256k1',
    'Invalid curve: Only secp256k1 is supported by BIP-32.',
  );

  return sharedDeriveChildKey(options, handleError);
}

/**
 * Handles an error thrown during derivation by incrementing the child index
 * and retrying.
 *
 * @param _ - The error that was thrown.
 * @param options - The options for deriving a child key.
 * @returns The options for deriving a child key with the child index
 * incremented by one.
 */
async function handleError(
  _: unknown,
  options: DeriveNodeArgs,
): Promise<DeriveNodeArgs> {
  const { childIndex, privateKey, publicKey, isHardened, curve, chainCode } =
    options;

  validateBIP32Index(childIndex + 1);

  if (privateKey) {
    const secretExtension = await deriveSecretExtension({
      privateKey,
      childIndex: childIndex + 1,
      isHardened,
      curve,
    });

    const newEntropy = generateEntropy({
      chainCode,
      extension: secretExtension,
    });

    return {
      ...options,
      childIndex: childIndex + 1,
      entropy: newEntropy,
    };
  }

  const publicExtension = derivePublicExtension({
    parentPublicKey: publicKey,
    childIndex: childIndex + 1,
  });

  const newEntropy = generateEntropy({
    chainCode,
    extension: publicExtension,
  });

  return {
    ...options,
    childIndex: childIndex + 1,
    entropy: newEntropy,
  };
}
