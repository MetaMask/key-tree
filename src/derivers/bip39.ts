import { mnemonicToSeed } from '@metamask/scure-bip39';
import { wordlist as englishWordlist } from '@metamask/scure-bip39/dist/wordlists/english';
import { hmac } from '@noble/hashes/hmac';
import { sha512 } from '@noble/hashes/sha512';

import { DeriveChildKeyArgs } from '.';
import { BIP39StringNode } from '../constants';
import { Curve, secp256k1 } from '../curves';
import { SLIP10Node } from '../SLIP10Node';
import { getFingerprint } from '../utils';

/**
 * Convert a BIP-39 mnemonic phrase to a multi path.
 *
 * @param mnemonic - The BIP-39 mnemonic phrase to convert.
 * @returns The multi path.
 */
export function bip39MnemonicToMultipath(mnemonic: string): BIP39StringNode {
  return `bip39:${mnemonic.toLowerCase().trim()}`;
}

/**
 * Create a {@link SLIP10Node} from a BIP-39 mnemonic phrase.
 *
 * @param options - The options for creating the node.
 * @param options.path - The multi path.
 * @param options.curve - The curve to use for derivation.
 * @returns The node.
 */
export async function deriveChildKey({
  path,
  curve,
}: DeriveChildKeyArgs): Promise<SLIP10Node> {
  return createBip39KeyFromSeed(
    await mnemonicToSeed(path, englishWordlist),
    curve,
  );
}

/**
 * Create a {@link SLIP10Node} from a BIP-39 seed.
 *
 * @param seed - The cryptographic seed bytes.
 * @param curve - The curve to use.
 * @returns An object containing the corresponding BIP-39 master key and chain
 * code.
 */
export async function createBip39KeyFromSeed(
  seed: Uint8Array,
  curve: Curve = secp256k1,
): Promise<SLIP10Node> {
  const key = hmac(sha512, curve.secret, seed);
  const privateKey = key.slice(0, 32);
  const chainCode = key.slice(32);

  const masterFingerprint = getFingerprint(
    await curve.getPublicKey(privateKey, true),
  );

  return SLIP10Node.fromExtendedKey({
    privateKey,
    chainCode,
    masterFingerprint,
    depth: 0,
    parentFingerprint: 0,
    index: 0,
    curve: curve.name,
  });
}
