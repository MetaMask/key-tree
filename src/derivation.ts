import {
  HDPathTuple,
  BIP44Depth,
  MAX_BIP_44_DEPTH,
  MIN_BIP_44_DEPTH,
  BIP_39_PATH_REGEX,
  BIP_32_PATH_REGEX,
} from './constants';
import { derivers, Deriver } from './derivers';

/**
 * ethereum default seed path: "m/44'/60'/0'/0/{account_index}"
 * multipath: "bip32:44'/bip32:60'/bip32:0'/bip32:0/bip32:{account_index}"
 *
 * m: { privateKey, chainCode } = sha512Hmac("Bitcoin seed", masterSeed)
 * 44': { privateKey, chainCode } = parentKey.privateKey + sha512Hmac(parentKey.chainCode, [0x00, parentKey.privateKey, index + HARDENED_OFFSET])
 * 60': { privateKey, chainCode } = parentKey.privateKey + sha512Hmac(parentKey.chainCode, [0x00, parentKey.privateKey, index + HARDENED_OFFSET])
 * 0': { privateKey, chainCode } = parentKey.privateKey + sha512Hmac(parentKey.chainCode, [0x00, parentKey.privateKey, index + HARDENED_OFFSET])
 * 0: { privateKey, chainCode } = parentKey.privateKey + sha512Hmac(parentKey.chainCode, [parentKey.publicKey, index])
 * 0: { privateKey, chainCode } = parentKey.privateKey + sha512Hmac(parentKey.chainCode, [parentKey.publicKey, index])
 */

/**
 * Takes a full or partial HD path string and returns the key corresponding to
 * the given path, with the following constraints:
 *
 * - If the path starts with a BIP-32 node, a parent key must be provided.
 * - If the path starts with a BIP-39 node, a parent key must NOT be provided.
 * - The path cannot exceed 5 BIP-32 nodes in length, optionally preceded by
 *   a single BIP-39 node.
 *
 * WARNING: It is the consumer's responsibility to ensure that the path is valid
 * relative to its parent key.
 *
 * @param pathSegment - A full or partial HD path, e.g.:
 * bip39:SEED_PHRASE/bip32:44'/bip32:60'/bip32:0'/bip32:0/bip32:0
 *
 * BIP-39 seed phrases must be lowercase, space-delimited, and 12-24 words long.
 * @param parentKey - The parent key of the given path segment, if any.
 * @returns The derived key.
 */
export function deriveKeyFromPath(
  pathSegment: HDPathTuple,
  parentKey?: Buffer,
  depth?: BIP44Depth,
): Buffer {
  if (parentKey && !Buffer.isBuffer(parentKey)) {
    throw new Error('Parent key must be a Buffer if specified.');
  }
  validatePathSegment(pathSegment, Boolean(parentKey), depth);

  let key = parentKey;

  // derive through each part of path
  pathSegment.forEach((node) => {
    const [pathType, pathValue] = node.split(':');
    /* istanbul ignore if: should be impossible */
    if (!hasDeriver(pathType)) {
      throw new Error(`Unknown derivation type: "${pathType}"`);
    }
    const deriver = derivers[pathType] as Deriver;
    const childKey = deriver.deriveChildKey(pathValue, key);
    // continue deriving from child key
    key = childKey;
  });

  return key as Buffer;
}

/**
 * @param pathType
 */
function hasDeriver(pathType: string): pathType is keyof typeof derivers {
  return pathType in derivers;
}

/**
 * The path segment must be one of the following:
 * - A lone BIP-32 path node
 * - A lone BIP-39 path node
 * - A multipath
 *
 * @param pathSegment - The path segment string to validate.
 */
export function validatePathSegment(
  pathSegment: HDPathTuple,
  hasKey: boolean,
  depth?: BIP44Depth,
) {
  if ((pathSegment as any).length === 0) {
    throw new Error(`Invalid HD path segment: The segment must not be empty.`);
  }

  if (pathSegment.length - 1 > MAX_BIP_44_DEPTH) {
    throw new Error(
      `Invalid HD path segment: The segment cannot exceed a 0-indexed depth of 5.`,
    );
  }

  let startsWithBip39 = false;
  pathSegment.forEach((node, index) => {
    if (index === 0) {
      startsWithBip39 = BIP_39_PATH_REGEX.test(node);
      if (!startsWithBip39 && !BIP_32_PATH_REGEX.test(node)) {
        throw getMalformedError();
      }
    } else if (!BIP_32_PATH_REGEX.test(node)) {
      throw getMalformedError();
    }
  });

  if (
    depth === MIN_BIP_44_DEPTH &&
    (!startsWithBip39 || pathSegment.length !== 1)
  ) {
    throw new Error(
      `Invalid HD path segment: The segment must consist of a single BIP-39 node for depths of ${MIN_BIP_44_DEPTH}. Received: "${pathSegment}"`,
    );
  }

  if (!hasKey && !startsWithBip39) {
    throw new Error(
      'Invalid derivation parameters: Must specify parent key if the first node of the path segment is not a BIP-39 node.',
    );
  }

  if (hasKey && startsWithBip39) {
    throw new Error(
      'Invalid derivation parameters: May not specify parent key if the path segment starts with a BIP-39 node.',
    );
  }
}

function getMalformedError() {
  throw new Error('Invalid HD path segment: The path segment is malformed.');
}
