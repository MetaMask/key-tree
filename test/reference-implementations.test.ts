import { BIP44Node } from '../src';
import { BIP44PurposeNodeToken, HDPathTuple } from '../src/constants';
import { ed25519 } from '../src/curves';
import { deriveKeyFromPath } from '../src/derivation';
import { privateKeyToEthAddress } from '../src/derivers/bip32';
import { createBip39KeyFromSeed } from '../src/derivers/bip39';
import {
  getBIP44CoinTypeToAddressPathTuple,
  hexStringToBuffer,
} from '../src/utils';
import { SLIP10Node } from '../src/SLIP10Node';
import fixtures from './fixtures';

describe('reference implementation tests', () => {
  describe('local', () => {
    const { addresses, mnemonic } = fixtures.local;
    const mnemonicBip39Node = `bip39:${mnemonic}` as const;

    describe('BIP44Node', () => {
      it('derives the expected keys', async () => {
        // Ethereum coin type node
        const node = await BIP44Node.create({
          derivationPath: [
            mnemonicBip39Node,
            BIP44PurposeNodeToken,
            `bip32:60'`,
          ],
        });

        for (let index = 0; index < addresses.length; index++) {
          const expectedAddress = addresses[index];
          const childNode = await node.derive(
            getBIP44CoinTypeToAddressPathTuple({ address_index: index }),
          );

          expect(await childNode.getAddress()).toStrictEqual(expectedAddress);
        }
      });
    });

    describe('deriveKeyFromPath', () => {
      it('derives the expected keys', async () => {
        // Ethereum coin type key
        const parentKey = await deriveKeyFromPath([
          mnemonicBip39Node,
          BIP44PurposeNodeToken,
          `bip32:60'`,
        ]);

        for (let index = 0; index < addresses.length; index++) {
          const expectedAddress = addresses[index];
          const childKey = await deriveKeyFromPath(
            getBIP44CoinTypeToAddressPathTuple({ address_index: index }),
            parentKey,
          );

          expect(
            privateKeyToEthAddress(childKey).toString('hex'),
          ).toStrictEqual(expectedAddress);
        }
      });
    });
  });

  describe('eth-hd-keyring', () => {
    const { mnemonic, addresses } = fixtures['eth-hd-keyring'];
    const mnemonicBip39Node = `bip39:${mnemonic}` as const;

    describe('BIP44Node', () => {
      it('derives the same keys as the reference implementation', async () => {
        // Ethereum coin type node
        const node = await BIP44Node.create({
          derivationPath: [
            mnemonicBip39Node,
            BIP44PurposeNodeToken,
            `bip32:60'`,
          ],
        });

        const numberOfAccounts = 5;
        const ourAccounts = [];
        for (let i = 0; i < numberOfAccounts; i++) {
          ourAccounts.push(
            await node
              .derive(getBIP44CoinTypeToAddressPathTuple({ address_index: i }))
              .then((childNode) => childNode.getAddress()),
          );
        }

        expect(addresses).toStrictEqual(
          ourAccounts.map((account) => `0x${account}`),
        );
      });
    });

    describe('deriveKeyFromPath', () => {
      it('derives the same keys as the reference implementation', async () => {
        // Ethereum coin type key
        const parentKey = await deriveKeyFromPath([
          mnemonicBip39Node,
          BIP44PurposeNodeToken,
          `bip32:60'`,
        ]);

        const numberOfAccounts = 5;
        const ourAccounts = [];
        for (let i = 0; i < numberOfAccounts; i++) {
          ourAccounts.push(
            privateKeyToEthAddress(
              await deriveKeyFromPath(
                getBIP44CoinTypeToAddressPathTuple({ address_index: i }),
                parentKey,
              ),
            ).toString('hex'),
          );
        }

        expect(addresses).toStrictEqual(
          ourAccounts.map((account) => `0x${account}`),
        );
      });
    });
  });

  describe('ethereumjs-wallet', () => {
    const { sampleAddressIndices, hexSeed, privateKey, address, path } =
      fixtures['ethereumjs-wallet'];
    const seed = hexStringToBuffer(hexSeed);

    describe('BIP44Node', () => {
      it('derives the same keys as the reference implementation', async () => {
        const seedKey = createBip39KeyFromSeed(seed);

        const parentNode = await BIP44Node.create({
          depth: 0,
          key: seedKey,
        });
        const node = await parentNode.derive(path.ours.tuple);

        expect(node.keyBuffer.slice(0, 32).toString('hex')).toStrictEqual(
          privateKey,
        );

        expect(await node.getAddress()).toStrictEqual(address);

        for (const { index, address: theirAddress } of sampleAddressIndices) {
          const ourAddress = privateKeyToEthAddress(
            (await node.derive([`bip32:${index}`])).keyBuffer,
          ).toString('hex');

          expect(ourAddress).toStrictEqual(theirAddress);
        }
      });
    });

    describe('deriveKeyFromPath', () => {
      it('derives the same keys as the reference implementation', async () => {
        const seedKey = createBip39KeyFromSeed(seed);
        const parentKey = await deriveKeyFromPath(path.ours.tuple, seedKey);

        expect(parentKey.slice(0, 32).toString('hex')).toStrictEqual(
          privateKey,
        );

        expect(privateKeyToEthAddress(parentKey).toString('hex')).toStrictEqual(
          address,
        );

        for (const { index, address: theirAddress } of sampleAddressIndices) {
          const ourAddress = privateKeyToEthAddress(
            await deriveKeyFromPath([`bip32:${index}`], parentKey),
          ).toString('hex');

          expect(ourAddress).toStrictEqual(theirAddress);
        }
      });
    });
  });

  describe('BIP-32 specification test vectors', () => {
    const vectors = fixtures.bip32;

    // We only test the BIP-32 vectors with deriveKeyFromPath, since not all
    // paths are BIP-44 compatible.
    describe('deriveKeyFromPath', () => {
      it('derives the test vector keys', async () => {
        for (const vector of vectors) {
          const seed = hexStringToBuffer(vector.hexSeed);
          const seedKey = createBip39KeyFromSeed(seed);

          for (const keyObj of vector.keys) {
            const { path, privateKey } = keyObj;

            let targetKey: Buffer;

            // If the path is empty, use the master node
            if (path.ours.string === '') {
              targetKey = seedKey;
            } else {
              targetKey = await deriveKeyFromPath(
                path.ours.tuple as HDPathTuple,
                seedKey,
              );
            }

            expect(targetKey.slice(0, 32).toString('hex')).toStrictEqual(
              privateKey,
            );
          }
        }
      });
    });
  });

  describe('ed25519', () => {
    describe('SLIP-10', () => {
      const vectors = fixtures.ed25519.slip10;

      describe('deriveKeyFromPath', () => {
        it('derives the test vector keys', async () => {
          for (const { hexSeed, keys } of vectors) {
            const seedKey = createBip39KeyFromSeed(
              hexStringToBuffer(hexSeed),
              ed25519,
            );

            for (const { path, privateKey, publicKey } of keys) {
              let targetKey: Buffer;
              if (path.ours.string === '') {
                targetKey = seedKey;
              } else {
                targetKey = await deriveKeyFromPath(
                  path.ours.tuple as HDPathTuple,
                  seedKey,
                  path.ours.tuple.length,
                  ed25519,
                );
              }

              expect(targetKey.slice(0, 32).toString('hex')).toBe(privateKey);
              expect(
                (await ed25519.getPublicKey(targetKey.slice(0, 32))).toString(
                  'hex',
                ),
              ).toBe(publicKey);
            }
          }
        });
      });
    });

    describe('ed25519-hd-key', () => {
      const { sampleKeyIndices, hexSeed, privateKey, path } =
        fixtures.ed25519['ed25519-hd-key'];
      const seed = hexStringToBuffer(hexSeed);

      describe('BIP44Node', () => {
        it('derives the same keys as the reference implementation', async () => {
          // Ethereum coin type node
          const seedKey = createBip39KeyFromSeed(seed, ed25519);

          const parentNode = await SLIP10Node.create({
            depth: 0,
            key: seedKey,
            curve: ed25519,
          });
          const node = await parentNode.derive(path.ours.tuple);

          expect(node.keyBuffer.slice(0, 32).toString('hex')).toStrictEqual(
            privateKey,
          );

          for (const {
            index,
            privateKey: theirPrivateKey,
            publicKey: theirPublicKey,
          } of sampleKeyIndices) {
            const childNode = await node.derive([`bip32:${index}'`]);
            const ourPrivateKey = childNode.keyBuffer.slice(0, 32);

            expect(ourPrivateKey.toString('hex')).toStrictEqual(
              theirPrivateKey,
            );

            expect(await childNode.getPublicKey()).toStrictEqual(
              theirPublicKey,
            );
          }
        });
      });

      describe('deriveKeyFromPath', () => {
        it('derives the same keys as the reference implementation', async () => {
          // Ethereum coin type key
          const seedKey = createBip39KeyFromSeed(seed, ed25519);
          const parentKey = await deriveKeyFromPath(
            [BIP44PurposeNodeToken, `bip32:0'`, `bip32:0'`, `bip32:1'`],
            seedKey,
            2,
            ed25519,
          );

          for (const {
            index,
            privateKey: theirPrivateKey,
          } of sampleKeyIndices) {
            const ourPrivateKey = await deriveKeyFromPath(
              [`bip32:${index}'`],
              parentKey,
              1,
              ed25519,
            );

            expect(ourPrivateKey.slice(0, 32).toString('hex')).toStrictEqual(
              theirPrivateKey,
            );
          }
        });
      });
    });
  });
});
