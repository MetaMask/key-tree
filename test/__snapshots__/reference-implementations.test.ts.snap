// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reference implementation tests eth-hd-keyring BIP44Node derives the same keys as the reference implementation 1`] = `
Array [
  "1c96099350f13d558464ec79b9be4445aa0ef579",
  "1b00aed43a693f3a957f9feb5cc08afa031e37a0",
  "8c9ba4f86ae12250ee1c3676ee925c77426d0b68",
  "ffe45dbc6c1bee8f211da2ec961f73b82e9ab42c",
  "b8a13c465c9a0a46f262a1ad666a752923e65b8c",
]
`;

exports[`reference implementation tests eth-hd-keyring deriveKeyFromPath derives the same keys as the reference implementation 1`] = `
Array [
  "1c96099350f13d558464ec79b9be4445aa0ef579",
  "1b00aed43a693f3a957f9feb5cc08afa031e37a0",
  "8c9ba4f86ae12250ee1c3676ee925c77426d0b68",
  "ffe45dbc6c1bee8f211da2ec961f73b82e9ab42c",
  "b8a13c465c9a0a46f262a1ad666a752923e65b8c",
]
`;

exports[`reference implementation tests ethereumjs-wallet BIP44Node derives the same keys as the reference implementation 1`] = `"c5ba325f997531b5f0f50868913f0ce2fc0386bd"`;

exports[`reference implementation tests ethereumjs-wallet BIP44Node derives the same keys as the reference implementation 2`] = `"326aa42f2b600f624d800e109b1e146906bc8175"`;

exports[`reference implementation tests ethereumjs-wallet BIP44Node derives the same keys as the reference implementation 3`] = `"6d0cc8671c91559d5f2d43de9c33eee0497db7cd"`;

exports[`reference implementation tests ethereumjs-wallet BIP44Node derives the same keys as the reference implementation 4`] = `"7bf971adda7f4487ac8f3dbd7450463ff3624f94"`;

exports[`reference implementation tests ethereumjs-wallet BIP44Node derives the same keys as the reference implementation 5`] = `"564d7507d39a881d04bffc0120ebd331e7c41758"`;

exports[`reference implementation tests ethereumjs-wallet BIP44Node derives the same keys as the reference implementation 6`] = `"7496ff062c1fe3e750dff9cbbe317558161ba6db"`;

exports[`reference implementation tests ethereumjs-wallet BIP44Node derives the same keys as the reference implementation 7`] = `"ddf1f1a72a668d5014ec57a24019291fd2a00197"`;

exports[`reference implementation tests ethereumjs-wallet BIP44Node derives the same keys as the reference implementation 8`] = `"24dd12b9df5375f3b7fc5a539d4ac1bd2c16e9a0"`;

exports[`reference implementation tests ethereumjs-wallet deriveKeyFromPath derives the same keys as the reference implementation 1`] = `"c5ba325f997531b5f0f50868913f0ce2fc0386bd"`;

exports[`reference implementation tests ethereumjs-wallet deriveKeyFromPath derives the same keys as the reference implementation 2`] = `"326aa42f2b600f624d800e109b1e146906bc8175"`;

exports[`reference implementation tests ethereumjs-wallet deriveKeyFromPath derives the same keys as the reference implementation 3`] = `"6d0cc8671c91559d5f2d43de9c33eee0497db7cd"`;

exports[`reference implementation tests ethereumjs-wallet deriveKeyFromPath derives the same keys as the reference implementation 4`] = `"7bf971adda7f4487ac8f3dbd7450463ff3624f94"`;

exports[`reference implementation tests ethereumjs-wallet deriveKeyFromPath derives the same keys as the reference implementation 5`] = `"564d7507d39a881d04bffc0120ebd331e7c41758"`;

exports[`reference implementation tests ethereumjs-wallet deriveKeyFromPath derives the same keys as the reference implementation 6`] = `"7496ff062c1fe3e750dff9cbbe317558161ba6db"`;

exports[`reference implementation tests ethereumjs-wallet deriveKeyFromPath derives the same keys as the reference implementation 7`] = `"ddf1f1a72a668d5014ec57a24019291fd2a00197"`;

exports[`reference implementation tests ethereumjs-wallet deriveKeyFromPath derives the same keys as the reference implementation 8`] = `"24dd12b9df5375f3b7fc5a539d4ac1bd2c16e9a0"`;
